// extras/arrayOfArrays.js
var aoa_vertical = function( pos ) {
    return this.map( function( el ) {
        return ( Array.isArray( el ) && el.length > pos ) ? el[pos] : null;
    });
};

var aoa_first = function( ) {
    return this.map( function( el ) {
        return ( Array.isArray( el ) && el.length > 0 ) ? el[0] : null;
    });
};

var aoa_last = function( ) {
    return this.map( function( el ) {
        return ( Array.isArray( el ) && el.length > 0 ) ? el[el.length - 1] : null;
    });
};

var aoa_longest = function( ) {

    var lgst = 0;

    for( var i = 0; i < this.length; i++ ) {
        if( Array.isArray( this[i] ) && this[i].length > lgst ) {
            lgst = this[i].length;
        }
    }

    return lgst;
};

var aoa_uniform = function( ) {

    var lens = { };

    for( var i = 0; i < this.length; i++ ) {

        if( Array.isArray( this[i] ) ) {
            lens[this[i].length.toString( )] = true
        } else {
            lens['-1'] = true;
        }
    }

    return ( Object.keys( lens ) === 1 );
};



// this causes trouble
module.exports.addAll = function( ) {
    Array.prototype.aoa_vertical = aoa_vertical;
    Array.prototype.aoa_first    = aoa_first;
    Array.prototype.aoa_last     = aoa_last;
    Array.prototype.aoa_longest  = aoa_longest;
    Array.prototype.aoa_uniform  = aoa_uniform;
};

// this shouldn't cause trouble outside of a
module.exports.bless = function( a ) {
    a.aoa_vertical = aoa_vertical.bind( a );
    a.aoa_first    = aoa_first.bind( a );
    a.aoa_last     = aoa_last.bind( a );
    a.aoa_longest  = aoa_longest.bind( a );
    a.aoa_uniform  = aoa_uniform.bind( a );
};
